src/puppetlabs/puppetdb/mq_listener.clj

```diff
@@ -339,25 +344,143 @@
-           (catch Throwable ex
-             (log/error ex "message receive failed")
-             (throw ex))))))
-    {:session sess :consumer consumer :producer producer}))
+(defn ^Session create-session [^Connection connection]
+  (.createSession connection false Session/CLIENT_ACKNOWLEDGE))
+
+(defn ^Queue create-queue [^Session session endpoint]
+  (.createQueue session endpoint))
+
+(defn ^Connection create-connection [^ConnectionFactory factory]
+  (doto (.createConnection factory)
+    (.setExceptionListener
+     (reify ExceptionListener
+       (onException [this ex]
+         (log/error ex "receiver queue connection error"))))
+    (.start)))
+
+(defn send-delayed-message [{:keys [^ConnectionFactory conn-pool endpoint]}]
+  (fn [message-to-convert message-properties]
+    (with-open [connection (create-connection conn-pool)
+                session (create-session connection)]
+      (let [q (create-queue session endpoint)]
+        (with-open [producer (.createProducer session q)]
+
+          (.send producer (mq/to-jms-message session message-to-convert message-properties)))))))
+
+(defn create-command-consumer
+  "Create and return a command handler. This function does the work of
+  consuming/storing a command. Handled commands are acknowledged here"
+  [{:keys [discard-dir] :as mq-context} command-handler]
+  (let [handle-message (wrap-message-handler-middleware (send-delayed-message mq-context)
+                                                        discard-dir
+                                                        command-handler)]
+    (fn [^Message message]
+      ;; When the queue is shutting down, it sends nil message
+      (when message
+        (try
+          (handle-message (mq/convert-jms-message message))
+          (.acknowledge message)
+          (catch Exception ex
+            (.rollback message)
+            (log/error ex "Unable to process message. Message not acknowledged and will be retried")))))))
+
+(defn create-mq-receiver
+  "Infinite loop, intended to be run in an isolated thread that
+  recieves a message from ActiveMQ, passes it to `mq-message-handler`
+  and looks for the next message"
+  [{:keys [^ConnectionFactory conn-pool endpoint]} mq-message-handler mq-connected-promise]
+  (with-open [connection (create-connection conn-pool)
+              session (create-session connection)
+              consumer (.createConsumer session (create-queue session endpoint))]
+    (try
+      ;; delivering the promise here we have ensured that we can
+      ;; connect to the MQ and that the queue has been created (for
+      ;; the first startup of PuppetDB)
+      (deliver mq-connected-promise true)
+      (loop []
+        (mq-message-handler (.receive consumer))
+        (recur))
+      (catch javax.jms.IllegalStateException e
+        (log/info "Received IllegalStateException, shutting down"))
+      (catch Exception e
+        (log/error e)))))
```

Comment:
Not sure, but wonder if it might also be worth an initial error message here (or in the caller) making it really clear that this receiver is dying.  Perhaps the exception is sufficient, but I could imagine that it might be quicker to grep for "receiver thread shutting down" or something.